### PR TITLE
Plugin to map feature

### LINKED ISSUE #10: Retrieve all field names and types
### Overview:
When editing a plugin, the frontend needs to know what inputs and input types are needed in order to prompt the user and validate any changes made

### Desired behavior:
Return a map of field to field type whenever the GET_PLUGIN request is handled

### PR DESCRIPTION
**Pull request by wangztim**
**Description of changes**

Added functionality to convert a plugin's struct to a map with reflection instead of json.(un)marshal

### CODE REVIEW

File: agent/agent.go [hunk 1]
```diff
@@ -330,32 +330,67 @@ func (a *Agent) GetPluginTypes(p interface{}) (map[string]interface{}, error) {
...
 		return data.Kind().String()
 	}
 }
 
+func (a *Agent) GetPluginValues(p interface{}) (map[string]interface{}, error) {
+	data := reflect.ValueOf(p).Elem() // extract Value of type interface{} from Value pointer to interface
+	values := getFieldValue(data)
+
+	v, ok := values.(map[string]interface{})
+
```
Comment (TheGoldenKyle) [inline agent/agent.go:363]: No space here (go convention)

File: agent/agent.go [hunk 2]
```diff
@@ -559,21 +594,38 @@ func updateStructValuesHelper(pluginPtr reflect.Value, newConfig map[string]inte
 	return plugin, map[string]interface{}{}, fmt.Errorf("could not update plugin")
 }
 
+func (a *Agent) CreateInput(name string) (interface{}, error) {
```
Comment (TheGoldenKyle) [inline agent/agent.go]: Why are we returning an ```interface{}``` for both of these? Also what are these creating? It just seems like it's running a plugin?

**Review by TheGoldenKyle — CHANGES_REQUESTED** [review]

File: agent/agent.go (hunk 2 above)
Comment (jodielu) [inline agent/agent.go]: Add comment for all exported functions (go convention)

File: agent/agent.go [hunk 3]
```diff
@@ -559,21 +594,38 @@ func updateStructValuesHelper(pluginPtr reflect.Value, newConfig map[string]inte
...
+	p, exists := inputs.Inputs[name]
+	if exists {
+		return p(), nil
+	}
+	return nil, fmt.Errorf("could not find input plugin with name: %s", name)
+}
+
+func (a *Agent) CreateOutput(name string) (interface{}, error) {
```
Comment (jodielu) [inline agent/agent.go]: Add comment for all exported functions (go convention)

File: assistant/assistant.go [hunk 4]
```diff
@@ -241,20 +241,28 @@ func (assistant *Assistant) getPlugin(req *request) response {
...
 	switch req.Plugin.Type {
 	case "INPUT":
-		plugin, err = assistant.agent.GetDefaultInputPlugin(req.Plugin.Name)
+		plugin, err = assistant.agent.CreateInput(req.Plugin.Name)
+		defaultValues, err = assistant.agent.GetDefaultInputPlugin(req.Plugin.Name)
 	case "OUTPUT":
-		plugin, err = assistant.agent.GetDefaultOutputPlugin(req.Plugin.Name)
+		plugin, err = assistant.agent.CreateOutput(req.Plugin.Name)
+		defaultValues, err = assistant.agent.GetDefaultOutputPlugin(req.Plugin.Name)
+	case "AGGREGATOR":
```
Comment (jodielu) [inline assistant/assistant.go:262]: Delete this case

File: assistant/assistant.go [hunk 5]
```diff
@@ -241,20 +241,28 @@ func (assistant *Assistant) getPlugin(req *request) response {
...
+		// TODO
+	case "PROCESSOR":
```
Comment (jodielu) [inline assistant/assistant.go]: Delete these cases

File: assistant/assistant_test.go [hunk 6]
```diff
@@ -72,12 +72,12 @@ func TestAssistant_GetInputPluginSchema(t *testing.T) {
 		"Duration": "int64",
 	}, m["ResponseTimeout"])
 
-// 	d, _ := json.Marshal(s.Defaults)
+	// 	d, _ := json.Marshal(s.Defaults)
```
Comment (jodielu) [inline assistant/assistant_test.go]: Why are these commented out?

**Review by jodielu — CHANGES_REQUESTED** [review]

File: assistant/assistant.go [hunk 7]
```diff
@@ -241,20 +241,28 @@ func (assistant *Assistant) getPlugin(req *request) response {
...
+
 	switch req.Plugin.Type {
 	case "INPUT":
-		plugin, err = assistant.agent.GetDefaultInputPlugin(req.Plugin.Name)
+		plugin, err = assistant.agent.CreateInput(req.Plugin.Name)
+		defaultValues, err = assistant.agent.GetDefaultInputPlugin(req.Plugin.Name)
 	case "OUTPUT":
-		plugin, err = assistant.agent.GetDefaultOutputPlugin(req.Plugin.Name)
+		plugin, err = assistant.agent.CreateOutput(req.Plugin.Name)
+		defaultValues, err = assistant.agent.GetDefaultOutputPlugin(req.Plugin.Name)
```
Comment (jodielu) [inline assistant/assistant.go]: Are these plugin variables used anywhere else? Why do we need to create these plugin objects

File: agent/agent.go [hunk 8]
```diff
@@ -667,7 +701,7 @@ func (a *Agent) UpdateOutputPlugin(name string, config map[string]interface{}) (
 			}
 
 			a.wg.Add(1)
-			a.StopInputPlugin(name, true)
+			a.StopOutputPlugin(name)
```
Comment (jodielu) [inline agent/agent.go:703]: Should this have a second parameter: true?

File: agent/agent.go [hunk 9]
```diff
@@ -330,32 +330,66 @@ func (a *Agent) GetPluginTypes(p interface{}) (map[string]interface{}, error) {
...
+	v, ok := values.(map[string]interface{})
+
+	if ok {
+		return v, nil
+	}
+
+	return nil, fmt.Errorf("returned schema is not a map")
+}
+
+// ? Should we make the nil check a parameter?
```
Comment (jodielu) [inline agent/agent.go]: Remove this comment - the function seems fine as is

**Comment by wangztim** [comment]
Merge?

**Pull request merged**